bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,348 @@
+      this.configuration = configuration;
+    }
+
+    @Override
+    public Coder<Result> getDefaultOutputCoder() {
+      return new HBaseResultCoder();
+    }
+
+    /**
+     * Splits the table based on keys that belong to "regions" a.k.a. tablets via the HBase
+     * {@link RegionLocator} interface which calls
+     * {@link BigtableService#sampleRowKeys(com.google.bigtable.v1.SampleRowKeysRequest, io.grpc.stub.StreamObserver)
+     * under the covers}.  Each BoundedResource will correspond to a single region.
+     */
+    @Override
+    public List<BoundedSource<Result>> splitIntoBundles(
+        long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+      TableName tableName = TableName.valueOf(configuration.getTable());
+      try (Connection conn = new BigtableConnection(configuration.toHBaseConfig());
+          Admin admin = conn.getAdmin(); RegionLocator locator = conn.getRegionLocator(tableName)) {
+        List<BoundedSource<Result>> splits = new ArrayList<>();
+        Pair<byte[][], byte[][]> startEndKeys = locator.getStartEndKeys();
+
+        byte[][] startKeys = startEndKeys.getFirst();
+        byte[][] endKeys = startEndKeys.getSecond();
+        for (int i = 0; i < startKeys.length; i++) {
+          splits.add(new SourceWithKeys(configuration, startKeys[i], endKeys[i]));
+        }
+        return splits;
+      }
+    }
+
+    @Override
+    public void validate() {
+      checkNotNullOrEmpty(configuration.getProjectId(), "project");
+      checkNotNullOrEmpty(configuration.getZone(), "zone");
+      checkNotNullOrEmpty(configuration.getCluster(), "cluster");
+      checkNotNullOrEmpty(configuration.getTable(), "table");
+    }
+
+    /*
+     * Right now, CBT doesn't have an easy way to export this information.
+     */
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+      // TODO(sduskis): Hook up the right size for the source.  The Cloud Bigtable API has the
+      // capability to get the size, but it's not currently exposed via HBase. (issue #400)
+      return 0;
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+      return true;
+    }
+
+    @Override
+    public BoundedSource.BoundedReader<Result> createReader(PipelineOptions options)
+        throws IOException {
+      return new Source.Reader(this, configuration, null, null);
+    }
+  }
+
+  /**
+   * Initialize the coders for the Bigtable Sink
+   *
+   * @return the original Pipeline.
+   */
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  public static Pipeline initializeForWrite(Pipeline p) {
+    // This enables the serialization of various Mutation types in the pipeline.
+    CoderRegistry registry = p.getCoderRegistry();
+    Coder coder = new HBaseMutationCoder();
+
+    // MutationCoder only supports Puts and Deletes. It will throw exceptions for Increment
```

Comment:
Do users set the timestamp by default? you don't do it in the example, which makes it not idempotent. I assume the idempotency requirement is because dataflow might retry. Do we know how often? Are users ok with seeing repeated entries at different timestamps?

At the very least this deserves a TODO.